src/api_server/src/lib.rs

```diff
@@ -317,37 +317,60 @@ impl ApiServer {
     }
 
     fn mmds_version(&self) -> Response {
-        let version = self
+        match self
             .mmds_info
             .lock()
             .expect("Failed to acquire lock on MMDS info")
-            .version();
-
-        ParsedRequest::success_response_with_data(&MmdsVersionType::from(version))
+            .as_ref()
+        {
+            Some(mmds) => {
+                let version = mmds.version();
+                ParsedRequest::success_response_with_data(&MmdsVersionType::from(version))
+            }
+            None => ApiServer::json_response(
+                StatusCode::BadRequest,
+                ApiServer::json_fault_message(MmdsError::NoMmds.to_string()),
+            ),
+        }
```

Comment:
While I agree the `mmds_unlock_and_do` call would look nice with a callback, I don't think the extra complexity added by a dynamic dispatched callback is worth the benefit here. If we really want to deduplicate code, I'd move towards a simpler, more modular solution where we replace just the MMDS unlocking part, which is actually the common denominator here. Moreover, the callback definition is not really extensible, because it has to take all the possible arguments for the "do" part, but in reality only one or none of them is used in a "do".

Reply:
Agree with you @georgepisaltu , although the part where we are returning the `NoMMDS` answer is also common, I think a macro with variable number of arguments (similar to println!) will work just right here. For now I am fine with returning just the unlocked part but we can try to simplify it afterwards.